### DISCUSSION TITLE
Why does the RocksDB example flush wal in apply_sm() vs append_entries()

### DISCUSSION THREAD

**Discussion by avantgardnerio** [databendlabs/openraft#616] [Q&A] [answered]
I've observed in the RocksDB example code, that the WAL is not flushed during [append_to_log()](https://github.com/datafuselabs/openraft/blob/c1c5f7553a466611a02c9096a51901f0c593e44f/examples/raft-kv-rocksdb/src/store.rs#L429).

It is, however, flushed during [apply_to_state_machine()](https://github.com/datafuselabs/openraft/blob/c1c5f7553a466611a02c9096a51901f0c593e44f/examples/raft-kv-rocksdb/src/store.rs#L510), which I found somewhat counter-intuitive.

Is it possible to explain (or point to docs) about what happens during a node crash & recovery? We are designing our persistent storage, and of course understanding these topics become very important to that design.

PS: I assume upon startup the first method called is `last_applied_state()` to see if the node is pristine or not? and if that state has an old raft log index as `last_applied` then the node will get all the intervening logs from it's peers and try to re-apply them? If that's true, why is it important to flush in `apply_to_state_machine()` if the logs will be re-applied anyway?

Sorry for all the questions, thanks for your help :raised_hands:

**Comment by drmingdrmer** [comment, marked as answer]
This issue and the closing PR #619 should have answered part of your question:
- #618

> PS: I assume upon startup the first method called is `last_applied_state()` to see if the node is pristine or not? 

Right.
The first step is to build a `RaftState` with the return value from `last_applied_state()` and other method:
https://github.com/datafuselabs/openraft/blob/cb2985ad6af8b0097e1842d4025f9532e1664fbd/openraft/src/raft_state.rs#L15

> and if that state has an old raft log index as last_applied then the node will get all the intervening logs from it's peers and try to re-apply them? 

Whether to apply these log has to be left to openraft to decide: If a leader has committed these log, the leader will send a committed-log-id to all nodes then every node calls `apply_to_state_machien()` with the logs that are smaller than the committed-log-id.

> If that's true, why is it important to flush in `apply_to_state_machine()` if the logs will be re-applied anyway?

It does not have to: 
- #620
